src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs

```diff
@@ -253,7 +264,86 @@ internal static unsafe byte[] UnsafeDecode(ReadOnlySpan<char> strSpan)
             }
 
             return result;
+        }
+
+        private static unsafe void UnsafeNetCoreDecode(ReadOnlySpan<char> strSpan, Span<byte> output, bool needReplace, int decodedLength)
+        {
+            // If the incoming chars don't contain any of the base64url characters that need to be replaced,
+            // and if the incoming chars are of the exact right length, then we'll be able to just pass the
+            // incoming chars directly to DecodeFromUtf8InPlace. Otherwise, rent an array, copy all the
+            // data into it, and do whatever fixups are necessary on that copy, then pass that copy into
+            // DecodeFromUtf8InPlace.
+
+            const int StackAllocThreshold = 512;
+            char[] arrayPoolChars = null;
+            scoped Span<char> charsSpan = default;
+            scoped ReadOnlySpan<char> source = strSpan;
+
+            if (needReplace || decodedLength != source.Length)
+            {
+                charsSpan = decodedLength <= StackAllocThreshold ?
+                    stackalloc char[StackAllocThreshold] :
+                    arrayPoolChars = ArrayPool<char>.Shared.Rent(decodedLength);
+                charsSpan = charsSpan.Slice(0, decodedLength);
+
+                source = HandlePaddingAndReplace(needReplace, charsSpan, source);
+            }
+
+            byte[] arrayPoolBytes = null;
+            Span<byte> bytesSpan = decodedLength <= StackAllocThreshold ?
+                stackalloc byte[StackAllocThreshold] :
+                arrayPoolBytes = ArrayPool<byte>.Shared.Rent(decodedLength);
+
+            int length = Encoding.UTF8.GetBytes(source, bytesSpan);
+            Span<byte> utf8Span = bytesSpan.Slice(0, length);
+            OperationStatus status = System.Buffers.Text.Base64.DecodeFromUtf8InPlace(utf8Span, out int bytesWritten);
+            Debug.Assert(status == OperationStatus.Done, "Expected DecodeFromUtf8 to be successful");
+
+            utf8Span.Slice(0, bytesWritten).CopyTo(output);
+
+            if (arrayPoolBytes is not null)
+            {
+                bytesSpan.Clear();
+                ArrayPool<byte>.Shared.Return(arrayPoolBytes);
+            }
+
+            if (arrayPoolChars is not null)
+            {
+                charsSpan.Clear();
+                ArrayPool<char>.Shared.Return(arrayPoolChars);
+            }
+        }
+
+        private static unsafe ReadOnlySpan<char> HandlePaddingAndReplace(bool needReplace, Span<char> charsSpan, ReadOnlySpan<char> source)
+        {
+            source.CopyTo(charsSpan);
```

Comment:
what happens if charsSpan is smaller than source?

Reply:
we allocate the buffer right before at the right size, we don't want this to be destructive